sjsonnet/src/sjsonnet/Std.scala

```diff
@@ -274,6 +274,40 @@ object Std {
         }
       )
     },
+    builtin("flatMap", "func", "arr"){ (ev, fs, func: Applyer, arr: Val) =>
+      val res: Val = arr match {
+        case a: Val.Arr =>
+          val arrs = Val.Arr(
+            a.value.map{ i =>
+              Val.Lazy(func.apply(i))
+            }
+          )
+          val out = collection.mutable.Buffer.empty[Val.Lazy]
+          for(x <- arrs.value){
+            x.force match{
+              case Val.Null => // do nothing
+              case Val.Arr(v) => out.appendAll(v)
+              case x => out.append(Val.Lazy(x))
+            }
+          }
+          Val.Arr(out.toSeq)
+
+        case s: Val.Str =>
+          val builder = new StringBuilder()
+          for (c: Char <- s.value) {
+            val fres = func.apply(Val.Lazy(Val.Str(c.toString)))
+            builder.append(
+              fres match {
+                case fstr: Val.Str => fstr.value
+                case x => Materializer(fres)(ev).toString
```

Comment:
1. We do not need this catch-all clause, since when the second parameter `arr` is a string, `func` returninganything except a string or `null` is an error

2. We *do* need a case to handle `null` here, since it appears that when the second parameter `arr` is a string, returning `null` means nothing gets added to the output string.

Reply:
Well technically nothing stops a user from implementing a function that returns something other than string. I think catch-all should throw an error saying "The response of the function parameter must be a string" and I will add a null check. What do you think?

Reply:
My proposed behavior above is according to what the upstream google/jsonnet behavior is, from what I can tell on jsonnet.org.

```jsonnet
> std.flatMap(function (x) 123, "helo")
RUNTIME ERROR: expected string but arr[0] was number
	std.jsonnet:271:7-77	function <anonymous>
	example1.jsonnet:1:1-38	
```

Although it's always possible to be a bit more accepting, my goal for this repo is to try and match google/jsonnet behavior as much as possible, including in the error cases on a best effort basis.

Reply:
Yes, agree completely, it is all fixed now and I added test for this type of error.